File: packages/adapters/src/http-adapter.ts
/**
 * The generic HTTP tool adapter.
 *
 * The transport is *injected*. There is no default, no fallback to
 * `globalThis.fetch` and no import of `node:http`, `node:https` or `node:net`
 * anywhere in this package, so a request plan can be constructed and asserted
 * with no network access at all.
 *
 * Every hop is authorized by `preflight` against the descriptor's network
 * profile: scheme, host and port allowlists (`P-014`..`P-016`), redirect
 * permission (`P-018`), redirect re-authorization (`P-017`) and the plaintext
 * downgrade refusal (`P-019`). Redirects are never followed by the transport —
 * the adapter follows them itself so that each hop is re-authorized.
 */

import {
  ok,
  isAborted,
  refused,
  runPreflight,
  type AdapterCallOptions,
  type AdapterOutcome,
  type AdapterResponse,
  type AdapterBaseOptions,
  type StreamingAdapter,
  type AdapterStream,
} from "./adapter.js";
import { CircuitBreaker } from "./circuit.js";
import { taxonomyFacts } from "./contract.js";
import { normalizedAdapterError } from "./envelope.js";
import { compareUnicodeCodePoints } from "./ordering.js";
import { composeUsage } from "./usage.js";
import type {
  AdapterCapability,
  AdapterDescriptor,
  AdapterErrorCode,
  AdapterRequest,
  AdapterUsage,
  EgressTarget,
  PreflightOutcome,
  ProviderMetricDeclaration,
  StreamFrame,
} from "./types.js";

export interface HttpRequestInit {
  readonly method: string;
  readonly headers: Readonly<Record<string, string>>;
  readonly body: string | null;
  /** Always manual: the adapter re-authorizes every redirect hop itself. */
  readonly redirect: "manual";
  readonly signal?: AbortSignal;
  /** Enforced by the injected transport; this package reads no clock. */
  readonly timeoutMs: number;
  readonly tlsMinimumVersion: "TLSv1.2" | "TLSv1.3";
}

export interface HttpResponseLike {
  readonly status: number;
  /** Case-insensitive single-value header lookup, as WHATWG `Headers` provides. */
  readonly headers: { get(name: string): string | null };
  readonly body?: string | null;
}

export type HttpFetch = (url: string, init: HttpRequestInit) => Promise<HttpResponseLike>;

export interface HttpRequestPlan {
  readonly target: EgressTarget;
  readonly url: string;
  readonly init: HttpRequestInit;
  readonly bodyBytes: number;
}

export interface HttpCallSpec {
  readonly method?: string;
  readonly path?: string;
  readonly headers?: Readonly<Record<string, string>>;
  readonly body?: string | null;
}

/**
 * The status-to-taxonomy map. Retryability is never read from a status text or
 * a payload: the status selects a code, and the code carries retryability.
 *
 * An unmapped 4xx is `GE_ADAPTER_INVALID_REQUEST` (the provider rejected the
 * request, provably no work). An unmapped 5xx is `GE_ADAPTER_TRANSPORT_FAILURE`
 * because the request was delivered and the effect is in doubt — the
 * conservative direction.
 */
export const DEFAULT_HTTP_STATUS_CODES: Readonly<Record<number, AdapterErrorCode>> =
  Object.freeze({
    400: "GE_ADAPTER_INVALID_REQUEST",
    401: "GE_ADAPTER_AUTHENTICATION",
    402: "GE_ADAPTER_QUOTA_EXCEEDED",
    403: "GE_ADAPTER_AUTHENTICATION",
    408: "GE_ADAPTER_TIMEOUT",
    422: "GE_ADAPTER_INVALID_REQUEST",
    429: "GE_ADAPTER_RATE_LIMITED",
    503: "GE_ADAPTER_RATE_LIMITED",
    504: "GE_ADAPTER_TIMEOUT",
  });

export function classifyHttpStatus(
  status: number,
  overrides: Readonly<Record<number, AdapterErrorCode>> = {},
): AdapterErrorCode | null {
  if (status >= 200 && status < 300) return null;
  const mapped = overrides[status] ?? DEFAULT_HTTP_STATUS_CODES[status];
  if (mapped !== undefined) return mapped;
  if (status >= 500) return "GE_ADAPTER_TRANSPORT_FAILURE";
  return "GE_ADAPTER_INVALID_REQUEST";
}

/**
 * `Retry-After` in delta-seconds only. An HTTP-date form is deliberately
 * ignored: converting it needs a wall clock, and this contract has none.
 */
export function parseRetryAfterMs(value: string | null): number | null {
  if (value === null) return null;
  if (!/^[0-9]{1,7}$/.test(value.trim())) return null;
  const ms = Number(value.trim()) * 1000;
  if (!Number.isSafeInteger(ms) || ms < 0 || ms > 3600000) return null;
  return ms;
}

export interface HttpAdapterOptions extends AdapterBaseOptions {
  /** Required. There is no default transport and no network fallback. */
  readonly fetch: HttpFetch;
  /**
   * Per-host credential headers. Credentials are host-isolated: a redirect to
   * another allowlisted host does not carry them, and they never appear in an
   * error message or a provider-safe detail field.
   */
  readonly credentials?: Readonly<Record<string, Readonly<Record<string, string>>>>;
  readonly statusCodes?: Readonly<Record<number, AdapterErrorCode>>;
  /** Bounded redirect chain. Defaults to 3 hops. */
  readonly maxRedirects?: number;
}

const DEFAULT_PORTS: Readonly<Record<string, number>> = Object.freeze({ http: 80, https: 443 });

function authority(target: EgressTarget): string {
  const standard = DEFAULT_PORTS[target.scheme];
  return standard === target.port ? target.host : `${target.host}:${String(target.port)}`;
}

function canonicalHeaders(headers: Readonly<Record<string, string>>): Record<string, string> {
  const entries = Object.entries(headers).map(
    ([name, value]) => [name.toLowerCase(), value] as const,
  );
  entries.sort((left, right) => compareUnicodeCodePoints(left[0], right[0]));
  return Object.fromEntries(entries);
}

export class HttpAdapter implements StreamingAdapter {
  readonly descriptor: AdapterDescriptor;
  readonly #fetch: HttpFetch;
  readonly #credentials: Readonly<Record<string, Readonly<Record<string, string>>>>;
  readonly #statusCodes: Readonly<Record<number, AdapterErrorCode>>;
  readonly #maxRedirects: number;
  readonly #policyMetrics: readonly ProviderMetricDeclaration[];
  readonly #forbiddenMarkers: readonly string[];
  readonly #breaker: CircuitBreaker;

  constructor(options: HttpAdapterOptions) {
    if (typeof options.fetch !== "function") {
      throw new Error("an HTTP adapter requires an injected fetch implementation");
    }
    this.descriptor = options.descriptor;
    this.#fetch = options.fetch;
    this.#credentials = options.credentials ?? {};
    this.#statusCodes = options.statusCodes ?? {};
    this.#maxRedirects = options.maxRedirects ?? 3;
    this.#policyMetrics = options.budgetPolicyAllowedProviderMetrics ?? [];
    this.#forbiddenMarkers = options.forbiddenMarkers ?? [];
    this.#breaker = new CircuitBreaker(options.descriptor.circuitPolicy);
  }

  capabilities(): readonly AdapterCapability[] {
    return this.descriptor.capabilities;
  }

  supports(capability: AdapterCapability): boolean {
    return this.descriptor.capabilities.includes(capability);
  }

  get circuit(): CircuitBreaker {
    return this.#breaker;
  }

  preflight(request: AdapterRequest): AdapterOutcome<PreflightOutcome> {
    return runPreflight(this.descriptor, request, this.#policyMetrics, 1, this.#forbiddenMarkers);
  }

  /**
   * Build the exact bytes that would be sent, without sending them. This is a
   * pure function of the descriptor, the request and the call spec, which is
   * what makes the corpus assertable with no transport at all.
   */
  buildRequestPlan(
    request: AdapterRequest,
    spec: HttpCallSpec = {},
    signal?: AbortSignal,
  ): HttpRequestPlan {
    const target = request.target;
    if (target === null) {
      throw new Error("an HTTP call requires a request target");
    }
    const path = spec.path ?? "/";
    const headers = canonicalHeaders({
      ...(spec.headers ?? {}),
      // Credentials are attached per host and never merged across hosts.
      ...(this.#credentials[target.host] ?? {}),
    });
    const body = spec.body ?? null;
    return Object.freeze({
      target,
      url: `${target.scheme}://${authority(target)}${path}`,
      bodyBytes: body === null ? 0 : body.length,
      init: Object.freeze({
        method: spec.method ?? "GET",
        headers: Object.freeze(headers),
        body,
        redirect: "manual" as const,
        timeoutMs: this.descriptor.bounds.requestTimeoutMs,
        tlsMinimumVersion: this.descriptor.network?.tlsMinimumVersion ?? "TLSv1.3",
        ...(signal === undefined ? {} : { signal }),
      }),
    });
  }

  async call(
    request: AdapterRequest,
    options: AdapterCallOptions = {},
    spec: HttpCallSpec = {},
  ): Promise<AdapterOutcome<AdapterResponse>> {
    const attempt = options.attempt ?? 1;
    const admitted = this.preflight(request);
    if (!admitted.ok) return refused(admitted.error);

    // adapter-semantics 8.3: cancellation before dispatch normalizes to nothing.
    if (isAborted(options.signal)) throw options.signal?.reason;

    let current = request;
    let currentSpec = spec;
    let hops = 0;
    let transportBytes = 0;
    let providerCalls = 0;

    for (;;) {
      const plan = this.buildRequestPlan(current, currentSpec, options.signal);
      transportBytes += plan.bodyBytes;
      providerCalls += 1;

      let response: HttpResponseLike;
      try {
        response = await this.#fetch(plan.url, plan.init);
      } catch (error) {
        return refused(this.#transportFailure(request, attempt, error, options));
      }

      const bodyBytes = response.body === null || response.body === undefined
        ? 0
        : response.body.length;
      transportBytes += bodyBytes;

      if (response.status >= 300 && response.status < 400) {
        const location = response.headers.get("location");
        hops += 1;
        if (location === null || hops > this.#maxRedirects) {
          return refused(
            this.#dispatchFailure(request, attempt, "GE_ADAPTER_TRANSPORT_FAILURE",
              "the provider redirected without a usable location or beyond the hop bound",
              null, transportBytes, providerCalls),
          );
        }
        const redirected = this.#redirectTarget(plan.target, location);
        if (redirected === null) {
          return refused(
            this.#dispatchFailure(request, attempt, "GE_ADAPTER_INVALID_REQUEST",
              "the provider redirect location is not a target this adapter can authorize",
              null, transportBytes, providerCalls),
          );
        }
        current = { ...current, target: redirected.target };
        currentSpec = { ...currentSpec, path: redirected.path };
        // Every hop is re-authorized against the same allowlist as the original.
        const reauthorized = this.preflight(current);
        if (!reauthorized.ok) return refused(reauthorized.error);
        continue;
      }

      if (bodyBytes > this.descriptor.bounds.maxResponseBytes) {
        // adapter-semantics 6.1: a provider that exceeds the declared response
        // bound has already performed work, so this is a dispatch code, not
        // GE_ADAPTER_BOUNDS_EXCEEDED.
        return refused(
          this.#dispatchFailure(request, attempt, "GE_ADAPTER_MALFORMED_RESPONSE",
            "the provider response exceeds maxResponseBytes", null,
            transportBytes, providerCalls),
        );
      }

      const code = classifyHttpStatus(response.status, this.#statusCodes);
      if (code !== null) {
        const retryAfterMs = parseRetryAfterMs(response.headers.get("retry-after"));
        return refused(
          this.#dispatchFailure(request, attempt, code,
            `the provider returned HTTP ${String(response.status)}`, retryAfterMs,
            transportBytes, providerCalls),
        );
      }

      return ok(
        Object.freeze({
          requestId: request.requestId,
          providerRequestId: this.#providerRequestId(response),
          finishReason: "stop" as const,
          text: response.body ?? "",
          structuredOutput: null,
          toolCalls: Object.freeze([]),
          usage: this.#usage(request, transportBytes, providerCalls, this.#providerRequestId(response)),
          normalizedStream: null,
        }),
      );
    }
  }

  /**
   * The HTTP adapter is a request/response tool adapter. It declares
   * `streaming` for callers that need frame-shaped delivery, and delivers the
   * single completed response as one conforming frame sequence.
   */
  stream(request: AdapterRequest, options: AdapterCallOptions = {}): AdapterStream {
    const completion = this.call(request, options);
    const emit = async function* (): AsyncGenerator<StreamFrame, void, void> {
      const outcome = await completion;
      if (!outcome.ok) return;
      yield { sequence: 0, kind: "start", bytes: 16 };
      yield { sequence: 1, kind: "text-delta", bytes: outcome.value.text.length };
      yield { sequence: 2, kind: "finish", bytes: 8, finishReason: "stop" };
    };
    return Object.freeze({ frames: emit(), completion });
  }

  #providerRequestId(response: HttpResponseLike): string | null {
    if (!this.supports("provider-request-id")) return null;
    const value = response.headers.get("x-request-id");
    return value === null || value.length === 0 ? null : value.slice(0, 256);
  }

  /**
   * The next hop a `Location` header names. Only absolute http(s) targets and
   * root-relative paths are representable; anything else is refused rather than
   * guessed, because a guess would be an unauthorized egress decision.
   */
  #redirectTarget(
    from: EgressTarget,
    location: string,
  ): { readonly target: EgressTarget; readonly path: string } | null {
    const absolute = /^(https?):\/\/([a-z0-9.-]+)(?::([0-9]{1,5}))?(\/[^\s]*)?$/i.exec(location);
    if (absolute !== null) {
      const scheme = (absolute[1] as string).toLowerCase();
      const host = (absolute[2] as string).toLowerCase();
      const port =
        absolute[3] === undefined ? (DEFAULT_PORTS[scheme] as number) : Number(absolute[3]);
      return Object.freeze({
        target: Object.freeze({ scheme, host, port, redirected: true, reauthorized: true }),
        path: absolute[4] ?? "/",
      });
    }
    if (location.startsWith("/")) {
      return Object.freeze({
        target: Object.freeze({ ...from, redirected: true, reauthorized: true }),
        path: location,
      });
    }
    return null;
  }

  #usage(
    request: AdapterRequest,
    transportBytes: number,
    providerCalls: number,
    providerRequestId: string | null,
  ): AdapterUsage {
    return composeUsage({
      descriptor: this.descriptor,
      requestId: request.requestId,
      providerRequestId,
      // An adapter without usage-reporting reports unknown trust and no
      // usage-unit meter (`U-018`). `transport-bytes` is a byte meter, so it
      // remains reportable.
      trust: this.supports("usage-reporting") ? "adapter-conservative" : "unknown",
      finishReason: null,
      meters: {
        "provider-calls": providerCalls,
        "transport-bytes": transportBytes,
      },
    });
  }

  #dispatchFailure(
    request: AdapterRequest,
    attempt: number,
    code: AdapterErrorCode,
    message: string,
    retryAfterMs: number | null,
    transportBytes: number,
    providerCalls: number,
  ) {
    const facts = taxonomyFacts(code);
    const conservative = facts.effectDisposition !== "not-applied";
    return normalizedAdapterError({
      descriptor: this.descriptor,
      requestId: request.requestId,
      attempt,
      code,
      sideEffectClass: request.sideEffectClass,
      message,
      retryAfterMs:
        retryAfterMs !== null && facts.retryable && this.supports("retry-after-hint")
          ? retryAfterMs
          : null,
      usage: conservative ? this.#usage(request, transportBytes, providerCalls, null) : null,
      forbiddenMarkers: this.#forbiddenMarkers,
    });
  }

  #transportFailure(
    request: AdapterRequest,
    attempt: number,
    error: unknown,
    options: AdapterCallOptions,
  ) {
    const name = error instanceof Error ? error.name : "";
    // The transport enforces the timeout; this package reads no clock.
    const code: AdapterErrorCode =
      name === "TimeoutError"
        ? "GE_ADAPTER_TIMEOUT"
        : isAborted(options.signal)
          ? "GE_ADAPTER_CANCELLED"
          : "GE_ADAPTER_TRANSPORT_FAILURE";
    // The cause is deliberately not quoted: a transport error can carry a URL
    // with credentials in it.
    return normalizedAdapterError({
      descriptor: this.descriptor,
      requestId: request.requestId,
      attempt,
      code,
      sideEffectClass: request.sideEffectClass,
      message: `the injected transport failed before a trusted response arrived (${name || "error"})`,
      usage: this.#usage(request, request.requestBytes, 1, null),
      forbiddenMarkers: this.#forbiddenMarkers,
    });
  }
}

export function createHttpAdapter(options: HttpAdapterOptions): HttpAdapter {
  return new HttpAdapter(options);
}

File: python/src/graph_engineering/adapters/http_adapter.py
"""The generic HTTP tool adapter.

The transport is *injected*.  There is no default, no fallback to any client
library and no import of ``http.client``, ``urllib.request``, ``socket`` or
``asyncio.open_connection`` anywhere in this package, so a request plan can be
constructed and asserted with no network access at all.

Every hop is authorized by :func:`preflight` against the descriptor's network
profile: scheme, host and port allowlists (``P-014``..``P-016``), redirect
permission (``P-018``), redirect re-authorization (``P-017``) and the plaintext
downgrade refusal (``P-019``).  Redirects are never followed by the transport —
the adapter follows them itself so that each hop is re-authorized.
"""

from __future__ import annotations

import asyncio
import re
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass, field, replace
from typing import Final

from .base import (
    AdapterCallOptions,
    AdapterOutcome,
    AdapterResponse,
    is_aborted,
    ok,
    refused,
    run_preflight,
)
from .circuit import CircuitBreaker
from .contract import taxonomy_facts
from .envelope import normalized_adapter_error
from .types import (
    AdapterCapability,
    AdapterDescriptor,
    AdapterErrorCode,
    AdapterErrorEnvelope,
    AdapterRequest,
    AdapterUsage,
    EgressTarget,
    PreflightOutcome,
    ProviderMetricDeclaration,
    StreamFrame,
)
from .usage import compose_usage

DEFAULT_PORTS: Final[dict[str, int]] = {"http": 80, "https": 443}
DEFAULT_MAX_REDIRECTS: Final = 3
_ABSOLUTE_LOCATION = re.compile(
    r"^(https?)://([A-Za-z0-9.-]+)(?::([0-9]{1,5}))?(/[^\s]*)?$", re.IGNORECASE
)
_DELTA_SECONDS = re.compile(r"^[0-9]{1,7}$")
_MAX_RETRY_AFTER_MS: Final = 3_600_000


@dataclass(frozen=True, slots=True)
class HttpRequestInit:
    """The exact bytes that would be sent, as a value."""

    method: str
    headers: Mapping[str, str]
    body: str | None
    #: Always manual: the adapter re-authorizes every redirect hop itself.
    redirect: str = "manual"
    #: Enforced by the injected transport; this package reads no clock.
    timeout_ms: int = 0
    tls_minimum_version: str = "TLSv1.3"


@dataclass(frozen=True, slots=True)
class HttpResponse:
    """What an injected transport returns.  There is no live counterpart."""

    status: int
    headers: Mapping[str, str] = field(default_factory=dict)
    body: str | None = None

    def header(self, name: str) -> str | None:
        wanted = name.lower()
        for key, value in self.headers.items():
            if key.lower() == wanted:
                return value
        return None


#: The injected transport.  Required; there is no module-level fallback.
HttpTransport = Callable[[str, HttpRequestInit], Awaitable[HttpResponse]]


class HttpTransportError(Exception):
    """A transport failed before a trusted response arrived."""


class HttpTransportTimeout(HttpTransportError):
    """The injected transport enforced the declared timeout."""


@dataclass(frozen=True, slots=True)
class HttpRequestPlan:
    target: EgressTarget
    url: str
    init: HttpRequestInit
    body_bytes: int


@dataclass(frozen=True, slots=True)
class HttpCallSpec:
    method: str = "GET"
    path: str = "/"
    headers: Mapping[str, str] = field(default_factory=dict)
    body: str | None = None


#: The status-to-taxonomy map.  Retryability is never read from a status text or
#: a payload: the status selects a code, and the code carries retryability.
#:
#: An unmapped 4xx is ``GE_ADAPTER_INVALID_REQUEST`` (the provider rejected the
#: request, provably no work).  An unmapped 5xx is
#: ``GE_ADAPTER_TRANSPORT_FAILURE`` because the request was delivered and the
#: effect is in doubt — the conservative direction.
DEFAULT_HTTP_STATUS_CODES: Final[dict[int, AdapterErrorCode]] = {
    400: "GE_ADAPTER_INVALID_REQUEST",
    401: "GE_ADAPTER_AUTHENTICATION",
    402: "GE_ADAPTER_QUOTA_EXCEEDED",
    403: "GE_ADAPTER_AUTHENTICATION",
    408: "GE_ADAPTER_TIMEOUT",
    422: "GE_ADAPTER_INVALID_REQUEST",
    429: "GE_ADAPTER_RATE_LIMITED",
    503: "GE_ADAPTER_RATE_LIMITED",
    504: "GE_ADAPTER_TIMEOUT",
}


def classify_http_status(
    status: int,
    overrides: Mapping[int, AdapterErrorCode] | None = None,
) -> AdapterErrorCode | None:
    if 200 <= status < 300:
        return None
    if overrides is not None and status in overrides:
        return overrides[status]
    mapped = DEFAULT_HTTP_STATUS_CODES.get(status)
    if mapped is not None:
        return mapped
    if status >= 500:
        return "GE_ADAPTER_TRANSPORT_FAILURE"
    return "GE_ADAPTER_INVALID_REQUEST"


def parse_retry_after_ms(value: str | None) -> int | None:
    """``Retry-After`` in delta-seconds only.

    An HTTP-date form is deliberately ignored: converting it needs a wall clock,
    and this contract has none.
    """
    if value is None:
        return None
    trimmed = value.strip()
    if _DELTA_SECONDS.fullmatch(trimmed) is None:
        return None
    milliseconds = int(trimmed) * 1000
    if milliseconds < 0 or milliseconds > _MAX_RETRY_AFTER_MS:
        return None
    return milliseconds


def _authority(target: EgressTarget) -> str:
    if DEFAULT_PORTS.get(target.scheme) == target.port:
        return target.host
    return f"{target.host}:{target.port}"


def _canonical_headers(headers: Mapping[str, str]) -> dict[str, str]:
    return {name.lower(): value for name, value in sorted(headers.items(), key=_lowered)}


def _lowered(entry: tuple[str, str]) -> str:
    return entry[0].lower()


class HttpAdapter:
    """A request/response tool adapter over an injected transport."""

    __slots__ = (
        "_breaker",
        "_credentials",
        "_forbidden_markers",
        "_max_redirects",
        "_policy_metrics",
        "_status_codes",
        "_transport",
        "descriptor",
    )

    def __init__(
        self,
        descriptor: AdapterDescriptor,
        *,
        transport: HttpTransport,
        credentials: Mapping[str, Mapping[str, str]] | None = None,
        status_codes: Mapping[int, AdapterErrorCode] | None = None,
        max_redirects: int = DEFAULT_MAX_REDIRECTS,
        budget_policy_allowed_provider_metrics: Sequence[ProviderMetricDeclaration] = (),
        forbidden_markers: Sequence[str] = (),
    ) -> None:
        if not callable(transport):
            raise ValueError("an HTTP adapter requires an injected transport")
        self.descriptor = descriptor
        self._transport = transport
        self._credentials = dict(credentials or {})
        self._status_codes = dict(status_codes or {})
        self._max_redirects = max_redirects
        self._policy_metrics = tuple(budget_policy_allowed_provider_metrics)
        self._forbidden_markers = tuple(forbidden_markers)
        self._breaker = CircuitBreaker(descriptor.circuit_policy)

    def capabilities(self) -> tuple[str, ...]:
        return self.descriptor.capabilities

    def supports(self, capability: AdapterCapability) -> bool:
        return capability in self.descriptor.capabilities

    @property
    def circuit(self) -> CircuitBreaker:
        return self._breaker

    def preflight(self, request: AdapterRequest) -> AdapterOutcome[PreflightOutcome]:
        return run_preflight(
            self.descriptor, request, self._policy_metrics, 1, self._forbidden_markers
        )

    def build_request_plan(
        self,
        request: AdapterRequest,
        spec: HttpCallSpec | None = None,
    ) -> HttpRequestPlan:
        """Build the exact bytes that would be sent, without sending them.

        This is a pure function of the descriptor, the request and the call
        spec, which is what makes the corpus assertable with no transport at
        all.
        """
        call = spec if spec is not None else HttpCallSpec()
        target = request.target
        if target is None:
            raise ValueError("an HTTP call requires a request target")
        # Credentials are attached per host and never merged across hosts.
        merged = {**call.headers, **self._credentials.get(target.host, {})}
        body = call.body
        return HttpRequestPlan(
            target=target,
            url=f"{target.scheme}://{_authority(target)}{call.path}",
            body_bytes=0 if body is None else len(body),
            init=HttpRequestInit(
                method=call.method,
                headers=_canonical_headers(merged),
                body=body,
                redirect="manual",
                timeout_ms=self.descriptor.bounds.request_timeout_ms,
                tls_minimum_version=(
                    self.descriptor.network.tls_minimum_version
                    if self.descriptor.network is not None
                    else "TLSv1.3"
                ),
            ),
        )

    async def call(
        self,
        request: AdapterRequest,
        options: AdapterCallOptions | None = None,
        spec: HttpCallSpec | None = None,
    ) -> AdapterOutcome[AdapterResponse]:
        settings = options if options is not None else AdapterCallOptions()
        attempt = settings.attempt
        admitted = self.preflight(request)
        if not admitted.ok:
            return refused(admitted.error)

        # adapter-semantics 8.3: cancellation before dispatch normalizes to
        # nothing at all — there is no adapter error to report.
        if is_aborted(settings.signal):
            raise _cancelled_before_dispatch(request.request_id)

        current = request
        current_spec = spec if spec is not None else HttpCallSpec()
        hops = 0
        transport_bytes = 0
        provider_calls = 0

        while True:
            plan = self.build_request_plan(current, current_spec)
            transport_bytes += plan.body_bytes
            provider_calls += 1

            try:
                response = await self._transport(plan.url, plan.init)
            except Exception as error:  # normalized into a code; never quoted
                return refused(self._transport_failure(request, attempt, error, settings))

            body_bytes = 0 if response.body is None else len(response.body)
            transport_bytes += body_bytes

            if 300 <= response.status < 400:
                location = response.header("location")
                hops += 1
                if location is None or hops > self._max_redirects:
                    return refused(
                        self._dispatch_failure(
                            request,
                            attempt,
                            "GE_ADAPTER_TRANSPORT_FAILURE",
                            "the provider redirected without a usable location "
                            "or beyond the hop bound",
                            None,
                            transport_bytes,
                            provider_calls,
                        )
                    )
                redirected = _redirect_target(plan.target, location)
                if redirected is None:
                    return refused(
                        self._dispatch_failure(
                            request,
                            attempt,
                            "GE_ADAPTER_INVALID_REQUEST",
                            "the provider redirect location is not a target "
                            "this adapter can authorize",
                            None,
                            transport_bytes,
                            provider_calls,
                        )
                    )
                target, path = redirected
                current = replace(current, target=target)
                current_spec = replace(current_spec, path=path)
                # Every hop is re-authorized against the same allowlist.
                reauthorized = self.preflight(current)
                if not reauthorized.ok:
                    return refused(reauthorized.error)
                continue

            if body_bytes > self.descriptor.bounds.max_response_bytes:
                # adapter-semantics 6.1: a provider that exceeds the declared
                # response bound has already performed work, so this is a
                # dispatch code, not GE_ADAPTER_BOUNDS_EXCEEDED.
                return refused(
                    self._dispatch_failure(
                        request,
                        attempt,
                        "GE_ADAPTER_MALFORMED_RESPONSE",
                        "the provider response exceeds maxResponseBytes",
                        None,
                        transport_bytes,
                        provider_calls,
                    )
                )

            code = classify_http_status(response.status, self._status_codes)
            if code is not None:
                retry_after_ms = parse_retry_after_ms(response.header("retry-after"))
                return refused(
                    self._dispatch_failure(
                        request,
                        attempt,
                        code,
                        f"the provider returned HTTP {response.status}",
                        retry_after_ms,
                        transport_bytes,
                        provider_calls,
                    )
                )

            provider_request_id = self._provider_request_id(response)
            return ok(
                AdapterResponse(
                    request_id=request.request_id,
                    provider_request_id=provider_request_id,
                    finish_reason="stop",
                    text=response.body or "",
                    structured_output=None,
                    tool_calls=(),
                    usage=self._usage(
                        request, transport_bytes, provider_calls, provider_request_id
                    ),
                    normalized_stream=None,
                )
            )

    async def stream(
        self,
        request: AdapterRequest,
        options: AdapterCallOptions | None = None,
        spec: HttpCallSpec | None = None,
    ) -> tuple[AdapterOutcome[AdapterResponse], tuple[StreamFrame, ...]]:
        """Deliver one completed response as one conforming frame sequence.

        The HTTP adapter is a request/response tool adapter.  It declares
        ``streaming`` for callers that need frame-shaped delivery and does not
        pretend to have an incremental source.
        """
        outcome = await self.call(request, options, spec)
        if not outcome.ok:
            return outcome, ()
        frames = (
            StreamFrame(sequence=0, kind="start", bytes=16),
            StreamFrame(sequence=1, kind="text-delta", bytes=len(outcome.value.text)),
            StreamFrame(sequence=2, kind="finish", bytes=8, finish_reason="stop"),
        )
        return outcome, frames

    def _provider_request_id(self, response: HttpResponse) -> str | None:
        if not self.supports("provider-request-id"):
            return None
        value = response.header("x-request-id")
        if value is None or not value:
            return None
        return value[:256]

    def _usage(
        self,
        request: AdapterRequest,
        transport_bytes: int,
        provider_calls: int,
        provider_request_id: str | None,
    ) -> AdapterUsage:
        return compose_usage(
            self.descriptor,
            request_id=request.request_id,
            provider_request_id=provider_request_id,
            # An adapter without usage-reporting reports unknown trust and no
            # usage-unit meter (`U-018`).  `transport-bytes` is a byte meter, so
            # it remains reportable.
            trust="adapter-conservative" if self.supports("usage-reporting") else "unknown",
            finish_reason=None,
            meters={"provider-calls": provider_calls, "transport-bytes": transport_bytes},
        )

    def _dispatch_failure(
        self,
        request: AdapterRequest,
        attempt: int,
        code: AdapterErrorCode,
        message: str,
        retry_after_ms: int | None,
        transport_bytes: int,
        provider_calls: int,
    ) -> AdapterErrorEnvelope:
        facts = taxonomy_facts(code)
        conservative = facts.effect_disposition != "not-applied"
        hinted = (
            retry_after_ms is not None and facts.retryable and self.supports("retry-after-hint")
        )
        return normalized_adapter_error(
            self.descriptor,
            request_id=request.request_id,
            attempt=attempt,
            code=code,
            side_effect_class=request.side_effect_class,
            message=message,
            retry_after_ms=retry_after_ms if hinted else None,
            usage=(
                self._usage(request, transport_bytes, provider_calls, None)
                if conservative
                else None
            ),
            forbidden_markers=self._forbidden_markers,
        )

    def _transport_failure(
        self,
        request: AdapterRequest,
        attempt: int,
        error: Exception,
        options: AdapterCallOptions,
    ) -> AdapterErrorEnvelope:
        # The transport enforces the timeout; this package reads no clock.
        if isinstance(error, HttpTransportTimeout | TimeoutError):
            code: AdapterErrorCode = "GE_ADAPTER_TIMEOUT"
        elif is_aborted(options.signal):
            code = "GE_ADAPTER_CANCELLED"
        else:
            code = "GE_ADAPTER_TRANSPORT_FAILURE"
        # The cause is deliberately not quoted: a transport error can carry a
        # URL with credentials in it.  Only the exception's type name survives.
        return normalized_adapter_error(
            self.descriptor,
            request_id=request.request_id,
            attempt=attempt,
            code=code,
            side_effect_class=request.side_effect_class,
            message=(
                "the injected transport failed before a trusted response "
                f"arrived ({type(error).__name__})"
            ),
            usage=self._usage(request, request.request_bytes, 1, None),
            forbidden_markers=self._forbidden_markers,
        )


def _cancelled_before_dispatch(request_id: str) -> BaseException:
    return asyncio.CancelledError(f"request {request_id!r} was cancelled before dispatch")


def _redirect_target(
    origin: EgressTarget,
    location: str,
) -> tuple[EgressTarget, str] | None:
    """The next hop a ``Location`` header names.

    Only absolute http(s) targets and root-relative paths are representable;
    anything else is refused rather than guessed, because a guess would be an
    unauthorized egress decision.
    """
    absolute = _ABSOLUTE_LOCATION.match(location)
    if absolute is not None:
        scheme = absolute.group(1).lower()
        host = absolute.group(2).lower()
        port_text = absolute.group(3)
        port = DEFAULT_PORTS.get(scheme, 0) if port_text is None else int(port_text)
        return (
            EgressTarget(
                scheme=scheme, host=host, port=port, redirected=True, reauthorized=True
            ),
            absolute.group(4) or "/",
        )
    if location.startswith("/"):
        return replace(origin, redirected=True, reauthorized=True), location
    return None


def create_http_adapter(
    descriptor: AdapterDescriptor,
    *,
    transport: HttpTransport,
    credentials: Mapping[str, Mapping[str, str]] | None = None,
    status_codes: Mapping[int, AdapterErrorCode] | None = None,
    max_redirects: int = DEFAULT_MAX_REDIRECTS,
    budget_policy_allowed_provider_metrics: Sequence[ProviderMetricDeclaration] = (),
    forbidden_markers: Sequence[str] = (),
) -> HttpAdapter:
    return HttpAdapter(
        descriptor,
        transport=transport,
        credentials=credentials,
        status_codes=status_codes,
        max_redirects=max_redirects,
        budget_policy_allowed_provider_metrics=budget_policy_allowed_provider_metrics,
        forbidden_markers=forbidden_markers,
    )
